### PR TITLE
Automatically reset PCs based on var size

### PR DESCRIPTION
**Pull request by hewgreen**
This is required for imaging omic datasets where the gene number (vars) is likely to be very low. Passing PC higher than genes would not be good.

### CODE REVIEW

File: scanpy_scripts/lib/_pca.py [hunk 1]
```diff
@@ -3,12 +3,20 @@
...
 def pca(adata, key_added=None, export_embedding=None, **kwargs):
     """
     Wrapper function for sc.pp.pca, for supporting named slot
     """
+
+    # Check the number of components (use_pc) less than floor(number of genes)/2
+    if 'use_pc' in kwargs and kwargs['use_pc'] is not None:
+        kwargs['use_pc'] = min(math.floor(adata.n_vars/2), kwargs['use_pc'])
+
+# minimum of (<value calculated above>, n_pcs_supplied)
```
Comment (pinin4fjords) [inline scanpy_scripts/lib/_pca.py]: This looks misplaced

**Comment by nh3** [comment]
`use_pc` is not a command line option for `pca()`, I guess you meant `n_comps`?
There's also `n_pcs` for `neighbors()`. They do different things.

BTW, `sc.pp.neighbors()` does fail with informative message if supplied `n_pcs` exists the dimension of `X_pca`.

**Comment by pinin4fjords** [comment]
Thanks for reviewing @nh3 . We just don't want the PCA tool to fail when we happen to supply a matrix that invalidates the hard-coded number of PCs. Analagous to the variable genes fix the other day, it should just knock the number of PCs down to the maximum allowed value (perhaps with a warning).

**Comment by hewgreen** [comment]
Sounds good to me. Should I swap to 'n_comps'?

**Comment by pinin4fjords** [comment]
> Sounds good to me. Should I swap to 'n_comps'?

Yep- see API doc for the function being called here: https://icb-scanpy.readthedocs-hosted.com/en/stable/api/scanpy.tl.pca.html

**Comment by pinin4fjords** [comment]
> I was looking here
> 
> https://github.com/ebi-gene-expression-group/scanpy-scripts/blob/0340ceadc635778154cdfad3bbe0d12f89158942/scanpy_scripts/cmd_options.py#L114
> 
> I wondered why this was different to scanpy.

That's just setting up the options for the cli. It's a generic thing that could be used in multiple function calls, not necessarily named the same as what gets passed to the actual Scanpy function.

**Comment by pinin4fjords** [comment]
> I was looking here
> 
> https://github.com/ebi-gene-expression-group/scanpy-scripts/blob/0340ceadc635778154cdfad3bbe0d12f89158942/scanpy_scripts/cmd_options.py#L114
> 
> I wondered why this was different to scanpy.

https://github.com/ebi-gene-expression-group/scanpy-scripts/blob/0340ceadc635778154cdfad3bbe0d12f89158942/scanpy_scripts/cmd_options.py#L160

**Comment by pinin4fjords** [comment]
Is this acceptable @nh3 ? An an example, Matt had a matrix with 33 genes. The default of 50 causes a nasty error becuase 50 >= floor(30/2). Seems more graceful to reduce to the maximum possible.

**Comment by nh3** [comment]
I think the method of modifying `n_comps` is fine. But I don't quite understand why reduce to half as many dimensions rather than `n_var`. Is that some kind of heuristics?

When not specifying `use_rep` and `n_var` < 50, the next step `neighbors()` automatically uses the expression matrix over the PCs; if `use_rep` is specified to PCA, there's `n_pcs` that can be used to reduce the number of desired PCs. So I don't see the necessity of applying heuristics here. Just my two cents.

**Comment by pinin4fjords** [comment]
Thanks @nh3 - yes, that appears to be the heuristic required (not my own invention, honest!). The 33 genes in the above example required no more than 15 PCs- it really isn't `n_var`  that's required as far as I can see. 

The need to apply the heuristics is that the PCA tool will die with small matrices if a default of 50 is left in place, killing the workflow, and we can't run the workflow unsupervised. Just doesn't seem graceful behaviour, and it's just nice to be able to say "50 when you can, otherwise as many as you can". 

I actually experienced an issue with the neighours step- it doesn't like the sparse matrices that come from using the expression matrix (was going to be the subject of a future PR)- so we're forcing to use the PCs for now.

**Comment by nh3** [comment]
Have you tried more datasets, and values of `svd_solver`? The heuristics sounds quite strange to me. scanpy actually handles the situation when `n_comps` > `n_var` by reducing it to `n_var` - 1. If that still fails, I suggest filing an issue to those guys.

https://github.com/theislab/scanpy/blob/85acb6c8949d43d08a26437dceab4fa5db79e246/scanpy/preprocessing/_simple.py#L451

**Comment by pinin4fjords** [comment]
@hewgreen will have to have a go- I'm on annual leave for 2 weeks very soon ;-). Here's the error trace:

```
Traceback (most recent call last):
  File "/nfs/production3/ma/home/gxa_galaxy/_conda/envs/__scanpy-scripts@0.2.5.post1/bin/scanpy-run-pca", line 10, in <module>
    sys.exit(PCA_CMD())
  File "/nfs/production3/ma/home/gxa_galaxy/_conda/envs/__scanpy-scripts@0.2.5.post1/lib/python3.6/site-packages/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "/nfs/production3/ma/home/gxa_galaxy/_conda/envs/__scanpy-scripts@0.2.5.post1/lib/python3.6/site-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/nfs/production3/ma/home/gxa_galaxy/_conda/envs/__scanpy-scripts@0.2.5.post1/lib/python3.6/site-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/nfs/production3/ma/home/gxa_galaxy/_conda/envs/__scanpy-scripts@0.2.5.post1/lib/python3.6/site-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/nfs/production3/ma/home/gxa_galaxy/_conda/envs/__scanpy-scripts@0.2.5.post1/lib/python3.6/site-packages/scanpy_scripts/cmd_utils.py", line 40, in cmd
    func(adata, **kwargs)
  File "/nfs/production3/ma/home/gxa_galaxy/_conda/envs/__scanpy-scripts@0.2.5.post1/lib/python3.6/site-packages/scanpy_scripts/lib/_pca.py", line 27, in pca
    sc.pp.pca(adata, **kwargs)
  File "/nfs/production3/ma/home/gxa_galaxy/_conda/envs/__scanpy-scripts@0.2.5.post1/lib/python3.6/site-packages/scanpy/preprocessing/_simple.py", line 504, in pca
    X_pca = pca_.fit_transform(X)
  File "/nfs/production3/ma/home/gxa_galaxy/_conda/envs/__scanpy-scripts@0.2.5.post1/lib/python3.6/site-packages/sklearn/decomposition/truncated_svd.py", line 175, in fit_transform
    " got %d >= %d" % (k, n_features))
ValueError: n_components must be < n_features; got 32 >= 15
```
... and here's the input object:

[test.h5.zip](https://github.com/ebi-gene-expression-group/scanpy-scripts/files/4236950/test.h5.zip)

**Comment by pcm32** [comment]
Ahh, ths was merged on another PR after (#78).

**Pull request closed without merging**